packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs

```diff
@@ -111,6 +111,43 @@ pub struct PostgresStore<C> {
     client: C,
 }
 
+#[async_trait]
+impl<C: AsClient> Store for PostgresStore<C> {
+    type Transaction<'t>
+    where
+        C: 't,
+    = PostgresStore<tokio_postgres::Transaction<'t>>;
+
+    async fn transaction(&mut self) -> Result<Self::Transaction<'_>, StoreError> {
+        Ok(PostgresStore::new(
+            self.as_mut_client()
+                .transaction()
+                .await
+                .into_report()
+                .change_context(StoreError)?,
+        ))
+    }
+}
+
+#[async_trait]
+impl Transaction for PostgresStore<tokio_postgres::Transaction<'_>> {
+    async fn commit(self) -> Result<(), StoreError> {
```

Comment:
Is there a way for us to expose the client so we don't have to do this? As we could end up just redefining the whole `Transaction` interface.

(For example) I'm wondering how we'd call `Transaction::prepare`, or `Transaction::execute`.

Reply:
We can't just impl `AsClient` or something?

Reply:
That's not possible as `Store` does not know anything about `Client`. It's intended, that you cannot call `prepare` or `execute` on a transaction directly as some stores may not have those methods defined. Instead, you directly call the functions defined on the `Store`, e.g. `create_entity`.

`AsClient` is an implementation detail for `Postgres`, which abstracts over `GenericClient`.

Reply:
Hm, I have the feeling our store abstraction isn't quite as satisfactory as we'd like, and that there are Postgres semantics bleeding over. But perhaps we'll find out with future development. Your explanation makes sense to me here though, thank you.